src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/ModuleTokenResolver.cs

```diff
@@ -436,5 +438,123 @@ public DummyTypeInfo GetTypeFromSpecification(MetadataReader reader, ModuleToken
                 return DummyTypeInfo.Instance;
             }
         }
+
+        private ImmutableHashSet<MethodDesc> KnownMethods;
```

Comment:
Immutable collections are meant to be used for collections updated in lock-free way. They pay extra overhead for the lock-free nature.

If you just a have a hashtable that gets initialized once and then just used for lookups, regular HashSet is going to work just fine and be faster.